### ISSUE TITLE
NVIDIA/COSMOFLOW implementation , mxnet error when increasing batch size

### ISSUE THREAD

**Issue by karanveersingh5623** [mlcommons/hpc_results_v1.0#2]
Benchmark running on Dell R750 XA server with 4 X A100s (80GB) .
Below is the working batch size == 16 trace 
`root@7881f8df9636:/workspace/cosmoflow# numactl --physcpubind=0-15,64-79 -- python train.py --log-prefix 'run_220315140551795650176_{}_1.log' --data-root-dir /mnt --num-epochs 3 --target-mae 0.124 --base-lr 0.004 --initial-lr 0.001 --momentum 0.9 --weight-decay 0.0 --warmup-epochs 1 --lr-scheduler-epochs 16 32 --lr-scheduler-decays 0.25 0.125 --training-batch-size 16 --validation-batch-size 16 --training-samples -1 --validation-samples -1 --data-layout NDHWC --data-shard-multiplier 1 --dali-num-threads 6 --shard-type local --seed 23739 --grad-prediv-factor 1.0 --instances 1 --spatial-span 1 --load-checkpoint '' --save-checkpoint /results/checkpoint.data --apply-log-transform --shuffle --preshuffle --use-amp
Namespace(apply_log_transform=True, base_lr=0.004, config_file=None, cuda_profiler_range='', dali_num_threads=6, dali_use_mmap=False, data_layout='NDHWC', data_root_dir=PosixPath('/mnt'), data_shard_multiplier=1, dropout=0.5, grad_prediv_factor=1.0, initial_lr=0.001, instances=1, load_checkpoint='', log_prefix='run_220315140551795650176_{}_1.log', lr_scheduler_decays=[0.25, 0.125], lr_scheduler_epochs=[16, 32], momentum=0.9, num_epochs=3, preshuffle=True, prestage=False, profile=False, save_checkpoint='/results/checkpoint.data', seed=23739, shard_type='local', shuffle=True, spatial_span=1, static_loss_scale=16384, target_mae=0.124, training_batch_size=16, training_samples=-1, use_amp=True, use_fp16=False, validation_batch_size=16, validation_samples=-1, warmup_epochs=1, weight_decay=0.0)
:::MLLOG {"namespace": "", "time_ms": 1647415346630, "event_type": "POINT_IN_TIME", "key": "cache_clear", "value": null, "metadata": {"file": "train.py", "lineno": 134}}
:::MLLOG {"namespace": "", "time_ms": 1647415346673, "event_type": "INTERVAL_START", "key": "init_start", "value": null, "metadata": {"file": "train.py", "lineno": 135}}
:::MLLOG {"namespace": "", "time_ms": 1647415346673, "event_type": "POINT_IN_TIME", "key": "submission_benchmark", "value": "cosmoflow", "metadata": {"file": "train.py", "lineno": 137}}
:::MLLOG {"namespace": "", "time_ms": 1647415346673, "event_type": "POINT_IN_TIME", "key": "submission_org", "value": "NVIDIA", "metadata": {"file": "train.py", "lineno": 139}}
:::MLLOG {"namespace": "", "time_ms": 1647415346674, "event_type": "POINT_IN_TIME", "key": "submission_division", "value": "closed", "metadata": {"file": "train.py", "lineno": 140}}
:::MLLOG {"namespace": "", "time_ms": 1647415346674, "event_type": "POINT_IN_TIME", "key": "submission_status", "value": "onprem", "metadata": {"file": "train.py", "lineno": 141}}
:::MLLOG {"namespace": "", "time_ms": 1647415346674, "event_type": "POINT_IN_TIME", "key": "submission_platform", "value": "1xNVIDIA DGX A100", "metadata": {"file": "train.py", "lineno": 142}}
:::MLLOG {"namespace": "", "time_ms": 1647415346674, "event_type": "POINT_IN_TIME", "key": "number_of_nodes", "value": 1, "metadata": {"file": "train.py", "lineno": 145}}
:::MLLOG {"namespace": "", "time_ms": 1647415346674, "event_type": "POINT_IN_TIME", "key": "accelerators_per_node", "value": 1, "metadata": {"file": "train.py", "lineno": 146}}
[07:22:26] ../src/storage/storage.cc:199: Using Pooled (Naive) StorageManager for CPU
[07:22:29] ../src/storage/storage.cc:199: Using Pooled (Naive) StorageManager for GPU

:::MLLOG {"namespace": "", "time_ms": 1647415377491, "event_type": "POINT_IN_TIME", "key": "opt_weight_decay", "value": 0.0, "metadata": {"file": "train.py", "lineno": 165}}
:::MLLOG {"namespace": "", "time_ms": 1647415377492, "event_type": "POINT_IN_TIME", "key": "dropout", "value": 0.5, "metadata": {"file": "train.py", "lineno": 167}}
:::MLLOG {"namespace": "", "time_ms": 1647415377734, "event_type": "POINT_IN_TIME", "key": "global_batch_size", "value": 16, "metadata": {"file": "/workspace/cosmoflow/data.py", "lineno": 352}}
:::MLLOG {"namespace": "", "time_ms": 1647415377735, "event_type": "POINT_IN_TIME", "key": "train_samples", "value": 117557, "metadata": {"file": "/workspace/cosmoflow/data.py", "lineno": 354}}
:::MLLOG {"namespace": "", "time_ms": 1647415377735, "event_type": "POINT_IN_TIME", "key": "eval_samples", "value": 166, "metadata": {"file": "/workspace/cosmoflow/data.py", "lineno": 355}}
:::MLLOG {"namespace": "", "time_ms": 1647415377735, "event_type": "POINT_IN_TIME", "key": "opt_base_learning_rate", "value": 0.004, "metadata": {"file": "train.py", "lineno": 92}}
:::MLLOG {"namespace": "", "time_ms": 1647415377736, "event_type": "POINT_IN_TIME", "key": "opt_learning_rate_warmup_epochs", "value": 1, "metadata": {"file": "train.py", "lineno": 94}}
:::MLLOG {"namespace": "", "time_ms": 1647415377736, "event_type": "POINT_IN_TIME", "key": "opt_learning_rate_warmup_factor", "value": 4, "metadata": {"file": "train.py", "lineno": 96}}
:::MLLOG {"namespace": "", "time_ms": 1647415377736, "event_type": "POINT_IN_TIME", "key": "opt_learning_rate_decay_boundary_epochs", "value": [16, 32], "metadata": {"file": "train.py", "lineno": 98}}
:::MLLOG {"namespace": "", "time_ms": 1647415377736, "event_type": "POINT_IN_TIME", "key": "opt_learning_rate_decay_factor", "value": [0.25, 0.125], "metadata": {"file": "train.py", "lineno": 100}}
:::MLLOG {"namespace": "", "time_ms": 1647415377736, "event_type": "POINT_IN_TIME", "key": "opt_name", "value": "sgd", "metadata": {"file": "train.py", "lineno": 184}}
:::MLLOG {"namespace": "", "time_ms": 1647415377737, "event_type": "INTERVAL_END", "key": "init_stop", "value": null, "metadata": {"file": "/workspace/cosmoflow/utils.py", "lineno": 144}}
:::MLLOG {"namespace": "", "time_ms": 1647415377738, "event_type": "INTERVAL_START", "key": "run_start", "value": null, "metadata": {"file": "train.py", "lineno": 206}}
:::MLLOG {"namespace": "", "time_ms": 1647415377738, "event_type": "INTERVAL_START", "key": "staging_start", "value": null, "metadata": {"file": "/workspace/cosmoflow/data.py", "lineno": 359}}
:::MLLOG {"namespace": "", "time_ms": 1647415380101, "event_type": "INTERVAL_END", "key": "staging_stop", "value": null, "metadata": {"file": "/workspace/cosmoflow/data.py", "lineno": 362, "staging_duration": 2.3632540702819824}}
:::MLLOG {"namespace": "", "time_ms": 1647415380101, "event_type": "INTERVAL_START", "key": "epoch_start", "value": null, "metadata": {"file": "train.py", "lineno": 215, "epoch_num": 1}}
`

**Comment by karanveersingh5623** [comment]
Do we need to change anything in here "model.py"
```
def init(self, ctx, batch_size: int, 
             use_amp: bool, use_wd: bool, 
             dist_desc: utils.DistributedEnvDesc,
             checkpoint: Optional[str] = None,
             dtype: str = "float32"):

        self.model.initialize(init=DefaultInitializer(layout=self.layout),
                              ctx=ctx)

        input_shape = (batch_size, 4, 128, 128, 128) if self.layout == "NCDHW" \
                else (batch_size, 128, 128, 128, 4)
        random_batch = (mx.nd.random.uniform(shape=input_shape, 
                                            ctx=ctx,
                                            dtype=dtype),
                        mx.nd.random.uniform(shape=(batch_size, 4),
                                            ctx=ctx,
                                            dtype="float32")) 
```

**Comment by hXl3s** [comment]
Hi @karanveersingh5623 

The issue you saw here is that MXNet uses 32bits to index to tensor, which means maximum number of elements in calculation is 2147483648. For BS=32, one of the layers exceeds this limits and I am afraid there is nothing that can be done here :(

**Comment by karanveersingh5623** [comment]
Thanks for update @hXl3s

**Comment by hXl3s** [comment]
@karanveersingh5623 Cosmoflow is using MPI to wire up everything. As long as you manage to run MPI jobs (1 per GPU), it should run fine. It is agnostic if you are using horovodrun, mpiexec or slurm based MPI launch

**Comment by karanveersingh5623** [comment]
@hXl3s , I have another query on model training .

Divided the dataset in 8 folders /xx/{1..8}/{train, validation} and tried running model training on 4 GPUs , total 8 docker containers
Only 4 containers run , as core utilization goes to 100% but GPU memory is only 25% used . Please refer the screenshot.
Is there any way where we can use full memory of A-100 GPU (80GB) to push more data into it ? or is it the same issue that we discussed above ? 
I am trying to run 2 seperate model training containers on a single GPU , it is possible ?

> **NOTE -  This is not slurm or MPI based run , i am directly running the training script inside a container , so its 1-1 mapping
![image](https://user-images.githubusercontent.com/30921663/168713384-b89d6162-9008-43c6-bc62-b642d826f08a.png)

**Comment by hXl3s** [comment]
I am not sure if filling all memory is worth it. With any BS>4 model becomes heavy compute bounded and it is nearly impossible to use additional VRAM to improve performance. 

Training two models should be possible, but it will impact performance significantly, most likely much more that can by gain from "full memory utilization" for cosmoflow so I strongly discourage doing that :)

**Comment by karanveersingh5623** [comment]
@hXl3s , thanks for reply 
The reason for asking preprocessing scaling and model training more data push is that we are trying to swap more IOs to NVMe Lustre file system to project the use of fast storage in HPC pipeline , whitepaper purpose
